File: src/services/apiBibleService.ts
import { BibleVerse } from '../types';

export interface ApiBibleChapterData {
  id: string;
  bibleId: string;
  number: string;
  bookId: string;
  reference: string;
  content?: any;
  verseCount?: number;
  copyright?: string;
}

// Canonical 66 Books USFM Codes
export const USFM_BOOK_CODES: string[] = [
  'GEN', 'EXO', 'LEV', 'NUM', 'DEU', 'JOS', 'JDG', 'RUT', '1SA', '2SA',
  '1KI', '2KI', '1CH', '2CH', 'EZR', 'NEH', 'EST', 'JOB', 'PSA', 'PRO',
  'ECC', 'SNG', 'ISA', 'JER', 'LAM', 'EZK', 'DAN', 'HOS', 'JOL', 'AMO',
  'OBA', 'JON', 'MIC', 'NAM', 'HAB', 'ZEP', 'HAG', 'ZEC', 'MAL', 'MAT',
  'MRK', 'LUK', 'JHN', 'ACT', 'ROM', '1CO', '2CO', 'GAL', 'EPH', 'PHP',
  'COL', '1TH', '2TH', '1TI', '2TI', 'TIT', 'PHM', 'HEB', 'JAS', '1PE',
  '2PE', '1JN', '2JN', '3JN', 'JUD', 'REV'
];

export const KNOWN_API_BIBLE_VERSIONS: Record<string, { id: string; name: string; abbreviation: string }> = {
  nbla: {
    id: 'ce11b813f9a27e20-01',
    name: 'Nueva Biblia de las Américas',
    abbreviation: 'NBLA'
  },
  bes: {
    id: 'b32b9d1b64b4ef29-01',
    name: 'La Biblia en Español Sencillo',
    abbreviation: 'BES'
  },
  vbl: {
    id: '482ddd53705278cc-02',
    name: 'Versión Biblia Libre',
    abbreviation: 'VBL'
  },
  pddpt: {
    id: '48acedcf8595c754-01',
    name: 'Palabra de Dios para ti',
    abbreviation: 'PdDpt'
  },
  rvr09: {
    id: '592420522e16049f-01',
    name: 'Reina Valera 1909 (API.Bible)',
    abbreviation: 'RVR09'
  },
  rvr09_api: {
    id: '592420522e16049f-01',
    name: 'Reina Valera 1909 (API.Bible)',
    abbreviation: 'RVR09'
  },
  kjv: {
    id: 'de4e12af7f28f599-01',
    name: 'King James Version',
    abbreviation: 'KJV'
  },
  bsb: {
    id: 'bba9f40183526463-01',
    name: 'Berean Standard Bible',
    abbreviation: 'BSB'
  },
  // Legacy aliases mapped to the best equivalents
  rvr1960: {
    id: '592420522e16049f-01',
    name: 'Reina Valera 1909 (API.Bible)',
    abbreviation: 'RVR09'
  },
  nvi: {
    id: 'nvi',
    name: 'Nueva Versión Internacional (NVI)',
    abbreviation: 'NVI'
  },
};

/**
 * Fetches all bibles associated with the user's API.Bible account from the backend
 */
export async function fetchAccountBibles(): Promise<any> {
  try {
    const res = await fetch('/api/bible/account-bibles');
    if (!res.ok) return null;
    return await res.json();
  } catch {
    return null;
  }
}

/**
 * Resolves USFM Book Code from book ID or 1-based book number
 */
export function resolveBookCode(bookIdOrNumber: string | number): string {
  if (typeof bookIdOrNumber === 'number') {
    return USFM_BOOK_CODES[bookIdOrNumber - 1] || 'GEN';
  }
  const clean = bookIdOrNumber.toUpperCase().trim();
  if (USFM_BOOK_CODES.includes(clean)) return clean;
  const num = parseInt(clean, 10);
  if (!isNaN(num) && num >= 1 && num <= 66) {
    return USFM_BOOK_CODES[num - 1] || 'GEN';
  }
  return clean.slice(0, 3) || 'GEN';
}

/**
 * Parses API.Bible response content (supports JSON AST, HTML strings, or plain text)
 * into a typed array of BibleVerse.
 */
export function parseApiBibleContent(
  content: any,
  bookId: string,
  bookName: string,
  chapter: number
): BibleVerse[] {
  const verses: BibleVerse[] = [];
  if (!content) return verses;

  // 1. AST JSON Array Structure (API.Bible JSON content-type)
  if (Array.isArray(content)) {
    let currentVerseNum = 0;
    let currentVerseText = '';

    const processItems = (items: any[]) => {
      for (const item of items) {
        if (!item) continue;
        if (item.name === 'verse' && item.attrs) {
          if (currentVerseNum > 0 && currentVerseText.trim()) {
            verses.push({
              bookId,
              bookName,
              chapter,
              verse: currentVerseNum,
              text: currentVerseText.replace(/\s+/g, ' ').trim()
            });
          }
          currentVerseNum = parseInt(item.attrs.number || item.attrs.sid?.split(':')[1] || '0', 10);
          currentVerseText = '';
        } else if (item.type === 'text' && typeof item.text === 'string') {
          currentVerseText += item.text;
        } else if (item.items && Array.isArray(item.items)) {
          processItems(item.items);
        }
      }
    };

    processItems(content);
    if (currentVerseNum > 0 && currentVerseText.trim()) {
      verses.push({
        bookId,
        bookName,
        chapter,
        verse: currentVerseNum,
        text: currentVerseText.replace(/\s+/g, ' ').trim()
      });
    }

    if (verses.length > 0) return verses;
  }

  // 2. String Content (HTML or Text)
  if (typeof content === 'string') {
    // Check if HTML with spans
    if (content.includes('<span') || content.includes('<p')) {
      try {
        // Strip out footnotes or extra titles
        const cleanHtml = content.replace(/<span class="note"[^>]*>.*?<\/span>/gi, '');
        // Regex to match verse spans like <span data-number="1" class="v">1</span> or <span class="v">1</span>
        const verseRegex = /<span[^>]*?(?:data-number="(\d+)"|class="v"[^>]*>(\d+))[^>]*>(?:(?:\d+)<\/span>)?([^<]+)/gi;
        let match;
        while ((match = verseRegex.exec(cleanHtml)) !== null) {
          const vNum = parseInt(match[1] || match[2], 10);
          const rawText = match[3] || '';
          if (vNum > 0 && rawText.trim()) {
            verses.push({
              bookId,
              bookName,
              chapter,
              verse: vNum,
              text: rawText.replace(/\s+/g, ' ').trim()
            });
          }
        }
      } catch (err) {
        console.warn('HTML regex parse fallback:', err);
      }
    }

    // If regex on HTML failed or plain text, parse bracketed verse numbers: [1] Texto... [2] Texto...
    if (verses.length === 0) {
      const textVerseRegex = /(?:\[(\d+)\]|(?:\b|^)(\d+)\s+)([\s\S]*?)(?=(?:\[\d+\]|(?:\b\d+\s+)|$))/g;
      let match;
      while ((match = textVerseRegex.exec(content)) !== null) {
        const vNum = parseInt(match[1] || match[2], 10);
        const text = (match[3] || '').replace(/<[^>]*>/g, '').replace(/\s+/g, ' ').trim();
        if (vNum > 0 && text) {
          verses.push({
            bookId,
            bookName,
            chapter,
            verse: vNum,
            text
          });
        }
      }
    }
  }

  return verses;
}

/**
 * Checks whether client has active internet connectivity
 */
export function isOnline(): boolean {
  return typeof navigator !== 'undefined' && typeof navigator.onLine === 'boolean'
    ? navigator.onLine
    : true;
}

/**
 * Fetches a chapter from API.Bible using the secure backend proxy (/api/bible/chapter)
 */
export async function fetchChapterFromApiBible(
  translationKey: string,
  bookIdOrNumber: string | number,
  chapter: number,
  bookName: string = '',
  customApiKey?: string
): Promise<{ verses: BibleVerse[]; isOfflineFallback?: boolean; fallbackMessage?: string }> {
  const cleanTr = translationKey.toLowerCase().trim();
  const bookCode = resolveBookCode(bookIdOrNumber);
  const chapterId = `${bookCode}.${chapter}`;
  const resolvedBibleId = KNOWN_API_BIBLE_VERSIONS[cleanTr]?.id || cleanTr;

  // Verify connectivity first
  if (!isOnline()) {
    throw new Error('OFFLINE_NO_INTERNET');
  }

  const url = `/api/bible/chapter?bibleId=${encodeURIComponent(resolvedBibleId)}&chapterId=${encodeURIComponent(chapterId)}&translation=${encodeURIComponent(cleanTr)}`;

  const headers: Record<string, string> = {
    Accept: 'application/json'
  };
  if (customApiKey) {
    headers['api-key'] = customApiKey;
  }

  const response = await fetch(url, { headers });
  if (!response.ok) {
    const errorBody = await response.json().catch(() => ({}));
    if (response.status === 401) {
      throw new Error(`API_BIBLE_UNAUTHORIZED: ${errorBody.message || 'Se requiere clave API.Bible válida.'}`);
    }
    throw new Error(errorBody.error || `Error ${response.status} al consultar API.Bible`);
  }

  const json = await response.json();
  const chapterData: ApiBibleChapterData = json.data;
  if (!chapterData) {
    throw new Error('Respuesta inválida de API.Bible');
  }

  const parsedVerses = parseApiBibleContent(
    chapterData.content,
    bookCode,
    bookName || chapterData.reference || bookCode,
    chapter
  );

  return { verses: parsedVerses };
}

File: flutter/lib/core/services/api_bible_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import '../constants/bible_translations.dart';
import '../../features/reader/data/services/local_bible_service.dart';

class ApiBibleService {
  static const String defaultBaseUrl = 'https://rest.api.bible/v1';

  // Configurable API key (can be passed via environment or settings)
  final String? apiKey;
  final String baseUrl;

  ApiBibleService({
    this.apiKey,
    this.baseUrl = defaultBaseUrl,
  });

  // Standard 66 Canonical Books USFM Codes
  static const List<String> usfmCodes = [
    'GEN', 'EXO', 'LEV', 'NUM', 'DEU', 'JOS', 'JDG', 'RUT', '1SA', '2SA',
    '1KI', '2KI', '1CH', '2CH', 'EZR', 'NEH', 'EST', 'JOB', 'PSA', 'PRO',
    'ECC', 'SNG', 'ISA', 'JER', 'LAM', 'EZK', 'DAN', 'HOS', 'JOL', 'AMO',
    'OBA', 'JON', 'MIC', 'NAM', 'HAB', 'ZEP', 'HAG', 'ZEC', 'MAL', 'MAT',
    'MRK', 'LUK', 'JHN', 'ACT', 'ROM', '1CO', '2CO', 'GAL', 'EPH', 'PHP',
    'COL', '1TH', '2TH', '1TI', '2TI', 'TIT', 'PHM', 'HEB', 'JAS', '1PE',
    '2PE', '1JN', '2JN', '3JN', 'JUD', 'REV'
  ];

  static String resolveBookCode(int bookNumber, [String? fallbackCode]) {
    if (bookNumber >= 1 && bookNumber <= usfmCodes.length) {
      return usfmCodes[bookNumber - 1];
    }
    if (fallbackCode != null && fallbackCode.isNotEmpty) {
      final clean = fallbackCode.toUpperCase().trim();
      if (usfmCodes.contains(clean)) return clean;
    }
    return 'GEN';
  }

  /// Checks if there is active internet connectivity
  Future<bool> checkInternetConnection() async {
    try {
      if (kIsWeb) {
        // On web, attempt a quick HEAD request or assume online
        return true;
      }
      final result = await InternetAddress.lookup('api.bible')
          .timeout(const Duration(seconds: 4));
      return result.isNotEmpty && result[0].rawAddress.isNotEmpty;
    } catch (_) {
      try {
        final result = await InternetAddress.lookup('google.com')
            .timeout(const Duration(seconds: 4));
        return result.isNotEmpty && result[0].rawAddress.isNotEmpty;
      } catch (_) {
        return false;
      }
    }
  }

  /// Fetches a chapter from API.Bible HTTP service
  Future<LocalBibleChapterResponse> fetchChapter({
    required String translationKey,
    required int bookNumber,
    required int chapterNumber,
    String? bookCode,
    String? bookName,
    String? activeApiKey,
  }) async {
    final config = BibleTranslationsCatalog.findById(translationKey);
    final bibleId = config.bibleId;

    if (bibleId == null || bibleId.isEmpty) {
      throw ArgumentError(
        'La versión "$translationKey" no tiene configurado un Bible ID para API.Bible.',
      );
    }

    final resolvedCode = resolveBookCode(bookNumber, bookCode);
    final chapterId = '$resolvedCode.$chapterNumber';

    final effectiveKey = activeApiKey ?? apiKey ?? '';
    final url = Uri.parse(
      '$baseUrl/bibles/$bibleId/chapters/$chapterId?content-type=json&include-verse-numbers=true&include-verse-spans=true',
    );

    final response = await http.get(
      url,
      headers: {
        'Accept': 'application/json',
        if (effectiveKey.isNotEmpty) 'api-key': effectiveKey,
      },
    ).timeout(const Duration(seconds: 12));

    if (response.statusCode != 200) {
      // If NVI returns 403 (unlicensed on standard API.Bible plan), fallback cleanly to NBLA
      if (response.statusCode == 403 && (translationKey == 'nvi' || bibleId == 'nvi')) {
        const nblaId = 'ce11b813f9a27e20-01';
        final fbUrl = Uri.parse(
          '$baseUrl/bibles/$nblaId/chapters/$chapterId?content-type=json&include-verse-numbers=true&include-verse-spans=true',
        );
        final fbResponse = await http.get(
          fbUrl,
          headers: {
            'Accept': 'application/json',
            if (effectiveKey.isNotEmpty) 'api-key': effectiveKey,
          },
        ).timeout(const Duration(seconds: 12));

        if (fbResponse.statusCode == 200) {
          final fbJsonMap = json.decode(utf8.decode(fbResponse.bodyBytes)) as Map<String, dynamic>;
          final fbData = fbJsonMap['data'] as Map<String, dynamic>?;
          if (fbData != null) {
            final rawContent = fbData['content'];
            final resolvedBookName = (fbData['reference'] as String?)?.split(' ')[0] ??
                bookName ??
                resolvedCode;
            final verses = _parseContent(
              rawContent,
              bookNumber,
              resolvedBookName,
              chapterNumber,
            );
            return LocalBibleChapterResponse(
              translation: 'Nueva Versión Internacional (NVI) [Equivalente NBLA]',
              abbreviation: 'NVI',
              bookNr: bookNumber,
              bookName: resolvedBookName,
              chapter: chapterNumber,
              name: '$resolvedBookName $chapterNumber',
              verses: verses,
            );
          }
        }
      }

      if (response.statusCode == 401) {
        throw StateError(
          'API.Bible requiere una clave válida (401 Unauthorized). '
          'Verifica la clave API o conéctate para descargar la versión.',
        );
      }
      throw StateError(
        'Error ${response.statusCode} al consultar API.Bible: ${response.body}',
      );
    }

    final jsonMap = json.decode(utf8.decode(response.bodyBytes)) as Map<String, dynamic>;
    final data = jsonMap['data'] as Map<String, dynamic>?;
    if (data == null) {
      throw StateError('Respuesta inesperada de API.Bible: formato inválido');
    }

    final rawContent = data['content'];
    final resolvedBookName = (data['reference'] as String?)?.split(' ')[0] ??
        bookName ??
        resolvedCode;

    final verses = _parseContent(
      rawContent,
      bookNumber,
      resolvedBookName,
      chapterNumber,
    );

    return LocalBibleChapterResponse(
      translation: config.name,
      abbreviation: config.abbreviation,
      bookNr: bookNumber,
      bookName: resolvedBookName,
      chapter: chapterNumber,
      name: '$resolvedBookName $chapterNumber',
      verses: verses,
    );
  }

  /// Robust parser supporting JSON AST, HTML spans, or formatted plain text from API.Bible
  List<LocalBibleVerseDto> _parseContent(
    dynamic content,
    int bookNr,
    String bookName,
    int chapter,
  ) {
    final List<LocalBibleVerseDto> verses = [];
    if (content == null) return verses;

    // 1. AST JSON list
    if (content is List) {
      int currentVerseNum = 0;
      final StringBuffer textBuf = StringBuffer();

      void traverse(List<dynamic> items) {
        for (final item in items) {
          if (item is! Map<String, dynamic>) continue;
          final name = item['name'];
          final attrs = item['attrs'];

          if (name == 'verse' && attrs is Map<String, dynamic>) {
            if (currentVerseNum > 0 && textBuf.isNotEmpty) {
              verses.add(LocalBibleVerseDto(
                chapter: chapter,
                verse: currentVerseNum,
                name: '$bookName $chapter:$currentVerseNum',
                text: textBuf.toString().replaceAll(RegExp(r'\s+'), ' ').trim(),
              ));
              textBuf.clear();
            }
            final rawNum = attrs['number'] ?? attrs['sid']?.toString().split(':').last;
            currentVerseNum = int.tryParse(rawNum?.toString() ?? '') ?? 0;
          } else if (item['type'] == 'text' && item['text'] is String) {
            textBuf.write(item['text']);
          } else if (item['items'] is List) {
            traverse(item['items'] as List<dynamic>);
          }
        }
      }

      traverse(content);
      if (currentVerseNum > 0 && textBuf.isNotEmpty) {
        verses.add(LocalBibleVerseDto(
          chapter: chapter,
          verse: currentVerseNum,
          name: '$bookName $chapter:$currentVerseNum',
          text: textBuf.toString().replaceAll(RegExp(r'\s+'), ' ').trim(),
        ));
      }

      if (verses.isNotEmpty) return verses;
    }

    // 2. String content (HTML or plain text)
    if (content is String) {
      // HTML format
      if (content.contains('<span') || content.contains('<p')) {
        final cleanHtml = content.replaceAll(RegExp(r'<span class="note"[^>]*>.*?</span>', dotAll: true), '');
        final regex = RegExp(r'<span[^>]*?(?:data-number="(\d+)"|class="v"[^>]*>(\d+))[^>]*>(?:(?:\d+)</span>)?([^<]+)');
        for (final match in regex.allMatches(cleanHtml)) {
          final numStr = match.group(1) ?? match.group(2) ?? '0';
          final text = match.group(3) ?? '';
          final vNum = int.tryParse(numStr) ?? 0;
          if (vNum > 0 && text.trim().isNotEmpty) {
            verses.add(LocalBibleVerseDto(
              chapter: chapter,
              verse: vNum,
              name: '$bookName $chapter:$vNum',
              text: text.replaceAll(RegExp(r'\s+'), ' ').trim(),
            ));
          }
        }
      }

      // Plain text fallback
      if (verses.isEmpty) {
        final textRegex = RegExp(r'(?:\[(\d+)\]|(?:\b|^)(\d+)\s+)([\s\S]*?)(?=(?:\[\d+\]|(?:\b\d+\s+)|$))');
        for (final match in textRegex.allMatches(content)) {
          final numStr = match.group(1) ?? match.group(2) ?? '0';
          final rawText = (match.group(3) ?? '')
              .replaceAll(RegExp(r'<[^>]*>'), '')
              .replaceAll(RegExp(r'\s+'), ' ')
              .trim();
          final vNum = int.tryParse(numStr) ?? 0;
          if (vNum > 0 && rawText.isNotEmpty) {
            verses.add(LocalBibleVerseDto(
              chapter: chapter,
              verse: vNum,
              name: '$bookName $chapter:$vNum',
              text: rawText,
            ));
          }
        }
      }
    }

    return verses;
  }
}
